{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This notebook runs the setup for JBrowse2 using Bash commands."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Instructions for Running the Notebook:\n",
    "\n",
    "1. Ensure you have created the APACHE_ROOT environment variable pointing to your Apache root directory.\n",
    "   Example: export APACHE_ROOT='/path/to/apache/root'\n",
    "   If you're unsure of the correct path, you can find it by running:\n",
    "   sudo find / -name \"www\" 2>/dev/null\n",
    "\n",
    "2. Ensure the required dependencies are installed, such as wget, samtools, bowtie2, and jbrowse.\n",
    "\n",
    "3. Modify the APACHE_ROOT and WORKDIR variables in the notebook to reflect your setup.\n",
    "\n",
    "4. Execute the notebook cells in order.\n",
    "\n",
    "Notes:\n",
    "- The notebook assumes you have a JBrowse2 instance installed in the $APACHE_ROOT/jbrowse2 directory.\n",
    "- If any errors occur, the notebook will output the error message and stop execution.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you have to restart due to errors, make sure to: \n",
    "1. Locate the www folder and clear jbrowse2 and any other contents in the folder (cd cd /usr/local/var/www) using rm -rf jbrowse2\n",
    "2. Next locate the tmp folder and then clear that as well \n",
    "3. Now run the following commands while in the tmp folder on your terminal "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#DO NOT RUN ON NOTEBOOK \n",
    "#clearing data terminal commands \n",
    "cd /usr/local/var/www\n",
    "rm -rf jbrowse2\n",
    "cd\n",
    "cd tmp \n",
    "rm -rf * #deletes everything in the tmp folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#DO NOT RUN ON NOTEBOOK \n",
    "# recreating the jbrowse folder, run while in tmp folder \n",
    "jbrowse create output_folder\n",
    "sudo mv output_folder $APACHE_ROOT/jbrowse2\n",
    "sudo chown -R $(whoami) $APACHE_ROOT/jbrowse2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Error Handling Function and Setting up Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working directory: /Users/smrithisurender/tmp\n",
      "Apache root directory: /usr/local/var/www\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "set -e  # Stop execution on errors\n",
    "\n",
    "# Define WORKDIR and APACHE_ROOT \n",
    "WORKDIR=\"/Users/smrithisurender/tmp\"  # Writable directory\n",
    "mkdir -p \"$WORKDIR\"\n",
    "chmod u+w \"$WORKDIR\"\n",
    "cd \"$WORKDIR\"\n",
    "\n",
    "APACHE_ROOT=\"/usr/local/var/www\"\n",
    "\n",
    "echo \"Working directory: $WORKDIR\"\n",
    "echo \"Apache root directory: $APACHE_ROOT\"\n",
    "\n",
    "# Ensure Apache root exists\n",
    "if [ ! -d \"$APACHE_ROOT/jbrowse2\" ]; then\n",
    "    echo \"Error: JBrowse2 not found at $APACHE_ROOT/jbrowse2. Please check your APACHE_ROOT variable.\"\n",
    "    exit 1\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download and Process Dengue Referencec Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Dengue genome (reference genome)...\n",
      "Decompressing reference genome file...\n",
      "Renaming reference genome file...\n",
      "Indexing reference genome file with samtools...\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "set -e  # Stop execution on errors\n",
    "\n",
    "cd \"/Users/smrithisurender/tmp\"  # Ensure you're in the working directory\n",
    "\n",
    "echo \"Downloading Dengue genome (reference genome)...\"\n",
    "wget -q \"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/862/125/GCF_000862125.1_ViralProj15306/GCF_000862125.1_ViralProj15306_genomic.fna.gz\" -O genome.fna.gz\n",
    "\n",
    "echo \"Decompressing reference genome file...\"\n",
    "gunzip -f genome.fna.gz || exit 1\n",
    "\n",
    "echo \"Renaming reference genome file...\"\n",
    "mv genome.fna viral_genome.fa || exit 1\n",
    "\n",
    "echo \"Indexing reference genome file with samtools...\"\n",
    "samtools faidx viral_genome.fa || exit 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Reference Genome to JBrowse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding reference genome assembly to JBrowse...\n",
      "Added assembly \"viral_genome\" to /usr/local/var/www/jbrowse2/config.json\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "set -e\n",
    "\n",
    "cd \"/Users/smrithisurender/tmp\"  \n",
    "\n",
    "APACHE_ROOT=\"/usr/local/var/www\"\n",
    "\n",
    "echo \"Adding reference genome assembly to JBrowse...\"\n",
    "jbrowse add-assembly viral_genome.fa --out \"$APACHE_ROOT/jbrowse2\" --load copy || exit 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modifying JBrowse2 configuration to add sequenceConfig and hide reverse strand...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Configuration updated successfully.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "set -e\n",
    "\n",
    "cd \"/Users/smrithisurender/tmp\"  \n",
    "\n",
    "APACHE_ROOT=\"/usr/local/var/www\"\n",
    "\n",
    "# Path to the JBrowse2 configuration file\n",
    "CONFIG_FILE=\"$APACHE_ROOT/jbrowse2/config.json\"\n",
    "\n",
    "echo \"Modifying JBrowse2 configuration to add sequenceConfig and hide reverse strand...\"\n",
    "\n",
    "# Add or update the sequenceConfig property\n",
    "jq '. + {sequenceConfig: {showReverseStrand: false}}' \"$CONFIG_FILE\" > \"$CONFIG_FILE.tmp\" && mv \"$CONFIG_FILE.tmp\" \"$CONFIG_FILE\"\n",
    "\n",
    "echo \"Configuration updated successfully.\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download and Process Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading genome annotations...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Decompressing annotations file...\n",
      "Sorting GFF3 annotations...\n",
      "Compressing sorted GFF3 file...\n",
      "Indexing compressed GFF3 file with tabix...\n",
      "Adding annotations track to JBrowse...\n",
      "Added track with name \"genes.gff\" and trackId \"genes.gff\" to /usr/local/var/www/jbrowse2/config.json\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "set -e\n",
    "\n",
    "cd \"/Users/smrithisurender/tmp\"  \n",
    "\n",
    "\n",
    "APACHE_ROOT=\"/opt/homebrew/var/www\"\n",
    "\n",
    "echo \"Downloading genome annotations...\"\n",
    "wget -q \"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/862/125/GCF_000862125.1_ViralProj15306/GCF_000862125.1_ViralProj15306_genomic.gff.gz\" -O annotations.gff.gz || exit 1\n",
    "\n",
    "echo \"Decompressing annotations file...\"\n",
    "gunzip -f annotations.gff.gz || exit 1\n",
    "\n",
    "echo \"Sorting GFF3 annotations...\"\n",
    "jbrowse sort-gff annotations.gff > genes.gff || exit 1\n",
    "\n",
    "echo \"Compressing sorted GFF3 file...\"\n",
    "bgzip -f genes.gff || exit 1\n",
    "\n",
    "echo \"Indexing compressed GFF3 file with tabix...\"\n",
    "tabix genes.gff.gz || exit 1\n",
    "\n",
    "echo \"Adding annotations track to JBrowse...\"\n",
    "jbrowse add-track genes.gff.gz --out \"$APACHE_ROOT/jbrowse2\" --load copy || exit 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Dengue genome (reference genome)...\n",
      "Decompressing reference genome file...\n",
      "Renaming reference genome file...\n",
      "Indexing reference genome file with samtools...\n",
      "Adding reference genome assembly to JBrowse...\n",
      "Added assembly \"dv_genome\" to /usr/local/var/www/jbrowse2/config.json\n",
      "Downloading genome annotations...\n",
      "Decompressing annotations file...\n",
      "Sorting GFF3 annotations...\n",
      "Compressing sorted GFF3 file...\n",
      "Indexing compressed GFF3 file with tabix...\n",
      "Adding annotations track to JBrowse...\n",
      "Added track with name \"dvgenes.gff\" and trackId \"dvgenes.gff\" to /usr/local/var/www/jbrowse2/config.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2024-11-25 15:38:18--  https://www.ncbi.nlm.nih.gov/sviewer/viewer.fcgi?id=OR486055.1&report=fasta&format=text\n",
      "Resolving www.ncbi.nlm.nih.gov (www.ncbi.nlm.nih.gov)... 2607:f220:41e:4290::110, 130.14.29.110\n",
      "Connecting to www.ncbi.nlm.nih.gov (www.ncbi.nlm.nih.gov)|2607:f220:41e:4290::110|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: unspecified [text/plain]\n",
      "Saving to: ‘dengue_virus1.fa’\n",
      "\n",
      "     0K ..........                                             31.8M=0s\n",
      "\n",
      "2024-11-25 15:38:19 (31.8 MB/s) - ‘dengue_virus1.fa’ saved [10954]\n",
      "\n",
      "--2024-11-25 15:38:19--  https://www.ncbi.nlm.nih.gov/sviewer/viewer.fcgi?id=OR771147.1&report=fasta&format=text\n",
      "Resolving www.ncbi.nlm.nih.gov (www.ncbi.nlm.nih.gov)... 2607:f220:41e:4290::110, 130.14.29.110\n",
      "Connecting to www.ncbi.nlm.nih.gov (www.ncbi.nlm.nih.gov)|2607:f220:41e:4290::110|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: unspecified [text/plain]\n",
      "Saving to: ‘dengue_virus2.fa’\n",
      "\n",
      "     0K ..........                                             37.1M=0s\n",
      "\n",
      "2024-11-25 15:38:19 (37.1 MB/s) - ‘dengue_virus2.fa’ saved [10813]\n",
      "\n",
      "--2024-11-25 15:38:19--  https://www.ncbi.nlm.nih.gov/sviewer/viewer.fcgi?id=OQ821525&report=fasta&format=text\n",
      "Resolving www.ncbi.nlm.nih.gov (www.ncbi.nlm.nih.gov)... 2607:f220:41e:4290::110, 130.14.29.110\n",
      "Connecting to www.ncbi.nlm.nih.gov (www.ncbi.nlm.nih.gov)|2607:f220:41e:4290::110|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: unspecified [text/plain]\n",
      "Saving to: ‘dengue_virus3.fa’\n",
      "\n",
      "     0K ..........                                             1.36M=0.008s\n",
      "\n",
      "2024-11-25 15:38:20 (1.36 MB/s) - ‘dengue_virus3.fa’ saved [10789]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "%%bash\n",
    "set -e \n",
    "\n",
    "\n",
    " # Stop execution on errors\n",
    "\n",
    "#Downloading all required files for Dengue Virus \n",
    "\n",
    "cd \"/Users/smrithisurender/tmp\"  # Ensure you're in the working directory\n",
    "APACHE_ROOT=\"/usr/local/var/www\"\n",
    "\n",
    "#getting the reference genome files \n",
    "echo \"Downloading Dengue genome (reference genome)...\"\n",
    "wget -q \"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/862/125/GCF_000862125.1_ViralProj15306/GCF_000862125.1_ViralProj15306_genomic.fna.gz\" -O dvgenome.fna.gz\n",
    "\n",
    "echo \"Decompressing reference genome file...\"\n",
    "gunzip -f dvgenome.fna.gz || exit 1\n",
    "\n",
    "echo \"Renaming reference genome file...\"\n",
    "mv dvgenome.fna dv_genome.fa || exit 1\n",
    "\n",
    "echo \"Indexing reference genome file with samtools...\"\n",
    "samtools faidx dv_genome.fa || exit 1\n",
    "\n",
    "echo \"Adding reference genome assembly to JBrowse...\"\n",
    "jbrowse add-assembly dv_genome.fa --out \"$APACHE_ROOT/jbrowse2\" --load copy || exit 1\n",
    "\n",
    "\n",
    "#getting the reference geneome annotaions \n",
    "echo \"Downloading genome annotations...\"\n",
    "wget -q \"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/862/125/GCF_000862125.1_ViralProj15306/GCF_000862125.1_ViralProj15306_genomic.gff.gz\" -O dvannotations.gff.gz || exit 1\n",
    "\n",
    "echo \"Decompressing annotations file...\"\n",
    "gunzip -f dvannotations.gff.gz || exit 1\n",
    "\n",
    "echo \"Sorting GFF3 annotations...\"\n",
    "jbrowse sort-gff dvannotations.gff > dvgenes.gff || exit 1\n",
    "\n",
    "echo \"Compressing sorted GFF3 file...\"\n",
    "bgzip -f dvgenes.gff || exit 1\n",
    "\n",
    "echo \"Indexing compressed GFF3 file with tabix...\"\n",
    "tabix dvgenes.gff.gz || exit 1\n",
    "\n",
    "echo \"Adding annotations track to JBrowse...\"\n",
    "jbrowse add-track dvgenes.gff.gz --out \"$APACHE_ROOT/jbrowse2\" --load copy || exit 1\n",
    "\n",
    "\n",
    "# getting the fasta files for the alignment \n",
    "# dengue virus 1\n",
    "wget \"https://www.ncbi.nlm.nih.gov/sviewer/viewer.fcgi?id=OR486055.1&report=fasta&format=text\" -O dengue_virus1.fa\n",
    "# dengue virus 2\n",
    "wget \"https://www.ncbi.nlm.nih.gov/sviewer/viewer.fcgi?id=OR771147.1&report=fasta&format=text\" -O dengue_virus2.fa\n",
    "#dengue virus 3\n",
    "wget \"https://www.ncbi.nlm.nih.gov/sviewer/viewer.fcgi?id=OQ821525&report=fasta&format=text\" -O dengue_virus3.fa\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/smrithisurender/bioe231fp/bowtie2-2.4.2-sra-linux-x86_64/bowtie2\n",
      "/usr/local/bin/samtools\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "which bowtie2\n",
    "which samtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing reference genome...\n",
      "Settings:\n",
      "  Output files: \"/Users/smrithisurender/tmp/dv_genome.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /Users/smrithisurender/tmp/dv_genome.fa\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 2683\n",
      "Using parameters --bmax 2013 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 2013 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 2, merged 7; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 2; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Avg bucket size: 1532.71 (target: 2012)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (2013) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 1449 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1450 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (2013) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 1668 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1669 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (2013) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 1970 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1971 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (2013) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 2005 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2006 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (2013) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 1766 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1767 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (2013) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 1080 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1081 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (2013) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 791 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 792 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 3426\n",
      "fchr[G]: 5666\n",
      "fchr[T]: 8436\n",
      "fchr[$]: 10735\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4198093 bytes to primary EBWT file: /Users/smrithisurender/tmp/dv_genome.1.bt2.tmp\n",
      "Wrote 2688 bytes to secondary EBWT file: /Users/smrithisurender/tmp/dv_genome.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 10735\n",
      "    bwtLen: 10736\n",
      "    sz: 2684\n",
      "    bwtSz: 2684\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 671\n",
      "    offsSz: 2684\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 56\n",
      "    numLines: 56\n",
      "    ebwtTotLen: 3584\n",
      "    ebwtTotSz: 3584\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 2683\n",
      "Using parameters --bmax 2013 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 2013 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 1532.71 (target: 2012)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (2013) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 1082 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1083 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (2013) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 1956 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1957 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (2013) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 2009 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2010 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (2013) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 1882 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1883 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (2013) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 1056 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1057 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (2013) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 1079 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1080 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (2013) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 1665 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1666 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 3426\n",
      "fchr[G]: 5666\n",
      "fchr[T]: 8436\n",
      "fchr[$]: 10735\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4198093 bytes to primary EBWT file: /Users/smrithisurender/tmp/dv_genome.rev.1.bt2.tmp\n",
      "Wrote 2688 bytes to secondary EBWT file: /Users/smrithisurender/tmp/dv_genome.rev.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 10735\n",
      "    bwtLen: 10736\n",
      "    sz: 2684\n",
      "    bwtSz: 2684\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 671\n",
      "    offsSz: 2684\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 56\n",
      "    numLines: 56\n",
      "    ebwtTotLen: 3584\n",
      "    ebwtTotSz: 3584\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Renaming /Users/smrithisurender/tmp/dv_genome.3.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.3.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.4.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.4.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.1.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.1.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.2.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.2.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.rev.1.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.rev.1.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.rev.2.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.rev.2.bt2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aligning comparison genome to dengue virus 1 genome using Bowtie2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwrote track with name \"dengue_virus1.sorted\" and trackId \"dengue_virus1.sorted\" in /usr/local/var/www/jbrowse2/config.json\n",
      "Settings:\n",
      "  Output files: \"/Users/smrithisurender/tmp/dv_genome.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /Users/smrithisurender/tmp/dv_genome.fa\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 2683\n",
      "Using parameters --bmax 2013 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 2013 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 2, merged 7; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 2; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Avg bucket size: 1532.71 (target: 2012)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (2013) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 1449 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1450 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (2013) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 1668 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1669 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (2013) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 1970 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1971 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (2013) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 2005 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2006 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (2013) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 1766 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1767 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (2013) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 1080 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1081 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (2013) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 791 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 792 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 3426\n",
      "fchr[G]: 5666\n",
      "fchr[T]: 8436\n",
      "fchr[$]: 10735\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4198093 bytes to primary EBWT file: /Users/smrithisurender/tmp/dv_genome.1.bt2.tmp\n",
      "Wrote 2688 bytes to secondary EBWT file: /Users/smrithisurender/tmp/dv_genome.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 10735\n",
      "    bwtLen: 10736\n",
      "    sz: 2684\n",
      "    bwtSz: 2684\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 671\n",
      "    offsSz: 2684\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 56\n",
      "    numLines: 56\n",
      "    ebwtTotLen: 3584\n",
      "    ebwtTotSz: 3584\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 2683\n",
      "Using parameters --bmax 2013 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 2013 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 1532.71 (target: 2012)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (2013) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 1082 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1083 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (2013) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 1956 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1957 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (2013) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 2009 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2010 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (2013) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 1882 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1883 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (2013) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 1056 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1057 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (2013) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 1079 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1080 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (2013) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 1665 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1666 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 3426\n",
      "fchr[G]: 5666\n",
      "fchr[T]: 8436\n",
      "fchr[$]: 10735\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4198093 bytes to primary EBWT file: /Users/smrithisurender/tmp/dv_genome.rev.1.bt2.tmp\n",
      "Wrote 2688 bytes to secondary EBWT file: /Users/smrithisurender/tmp/dv_genome.rev.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 10735\n",
      "    bwtLen: 10736\n",
      "    sz: 2684\n",
      "    bwtSz: 2684\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 671\n",
      "    offsSz: 2684\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 56\n",
      "    numLines: 56\n",
      "    ebwtTotLen: 3584\n",
      "    ebwtTotSz: 3584\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:00\n",
      "Aligning comparison genome to dengue virus 2 genome using Bowtie2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Renaming /Users/smrithisurender/tmp/dv_genome.3.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.3.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.4.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.4.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.1.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.1.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.2.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.2.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.rev.1.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.rev.1.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.rev.2.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.rev.2.bt2\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    1 (100.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "0.00% overall alignment rate\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwrote track with name \"dengue_virus2.sorted\" and trackId \"dengue_virus2.sorted\" in /usr/local/var/www/jbrowse2/config.json\n",
      "Settings:\n",
      "  Output files: \"/Users/smrithisurender/tmp/dv_genome.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /Users/smrithisurender/tmp/dv_genome.fa\n",
      "Reading reference sizes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 2683\n",
      "Using parameters --bmax 2013 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 2013 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 2, merged 7; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 2; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Avg bucket size: 1532.71 (target: 2012)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (2013) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 1449 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1450 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (2013) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 1668 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1669 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (2013) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 1970 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1971 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (2013) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 2005 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2006 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (2013) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 1766 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1767 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (2013) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 1080 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1081 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (2013) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 791 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 792 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 3426\n",
      "fchr[G]: 5666\n",
      "fchr[T]: 8436\n",
      "fchr[$]: 10735\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4198093 bytes to primary EBWT file: /Users/smrithisurender/tmp/dv_genome.1.bt2.tmp\n",
      "Wrote 2688 bytes to secondary EBWT file: /Users/smrithisurender/tmp/dv_genome.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 10735\n",
      "    bwtLen: 10736\n",
      "    sz: 2684\n",
      "    bwtSz: 2684\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 671\n",
      "    offsSz: 2684\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 56\n",
      "    numLines: 56\n",
      "    ebwtTotLen: 3584\n",
      "    ebwtTotSz: 3584\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 2683\n",
      "Using parameters --bmax 2013 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 2013 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 1532.71 (target: 2012)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (2013) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 1082 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1083 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (2013) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 1956 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1957 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (2013) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 2009 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2010 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (2013) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 1882 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1883 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (2013) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 1056 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1057 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (2013) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 1079 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1080 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (2013) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 1665 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1666 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 3426\n",
      "fchr[G]: 5666\n",
      "fchr[T]: 8436\n",
      "fchr[$]: 10735\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4198093 bytes to primary EBWT file: /Users/smrithisurender/tmp/dv_genome.rev.1.bt2.tmp\n",
      "Wrote 2688 bytes to secondary EBWT file: /Users/smrithisurender/tmp/dv_genome.rev.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 10735\n",
      "    bwtLen: 10736\n",
      "    sz: 2684\n",
      "    bwtSz: 2684\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 671\n",
      "    offsSz: 2684\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 56\n",
      "    numLines: 56\n",
      "    ebwtTotLen: 3584\n",
      "    ebwtTotSz: 3584\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:00\n",
      "Aligning comparison genome to dengue virus 3 genome using Bowtie2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Renaming /Users/smrithisurender/tmp/dv_genome.3.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.3.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.4.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.4.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.1.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.1.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.2.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.2.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.rev.1.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.rev.1.bt2\n",
      "Renaming /Users/smrithisurender/tmp/dv_genome.rev.2.bt2.tmp to /Users/smrithisurender/tmp/dv_genome.rev.2.bt2\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    1 (100.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "0.00% overall alignment rate\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwrote track with name \"dengue_virus3.sorted\" and trackId \"dengue_virus3.sorted\" in /usr/local/var/www/jbrowse2/config.json\n",
      "Comparison of dengue virus alignments successfully added to JBrowse.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "set -e \n",
    "\n",
    "\n",
    " # Stop execution on errors\n",
    "# processing the reference genome and then creating alignments compared to the other genomes \n",
    "export PATH=$PATH:/usr/local/bin\n",
    "\n",
    "cd \"/Users/smrithisurender/tmp\"  # Ensure you're in the working directory\n",
    "APACHE_ROOT=\"/usr/local/var/www\"\n",
    "WORKDIR=\"/Users/smrithisurender/tmp\"\n",
    "# creating the dengue virus alignments \n",
    "echo \"Preparing reference genome...\"\n",
    "bowtie2-build $WORKDIR/dv_genome.fa $WORKDIR/dv_genome\n",
    "\n",
    "\n",
    "echo \"Aligning comparison genome to dengue virus 1 genome using Bowtie2...\"\n",
    "bowtie2 -x $WORKDIR/dv_genome -f $WORKDIR/dengue_virus1.fa -S $WORKDIR/dengue_virus1.sam --very-sensitive\n",
    "samtools view -bS $WORKDIR/dengue_virus1.sam > $WORKDIR/dengue_virus1.bam\n",
    "samtools sort $WORKDIR/dengue_virus1.bam -o $WORKDIR/dengue_virus1.sorted.bam\n",
    "samtools index $WORKDIR/dengue_virus1.sorted.bam\n",
    "\n",
    "jbrowse add-track $WORKDIR/dengue_virus1.sorted.bam --out $APACHE_ROOT/jbrowse2 --load copy --force\n",
    "\n",
    "bowtie2-build $WORKDIR/dv_genome.fa $WORKDIR/dv_genome\n",
    "echo \"Aligning comparison genome to dengue virus 2 genome using Bowtie2...\"\n",
    "bowtie2 -x $WORKDIR/dv_genome -f $WORKDIR/dengue_virus2.fa -S $WORKDIR/dengue_virus2.sam --very-sensitive\n",
    "samtools view -bS $WORKDIR/dengue_virus2.sam > $WORKDIR/dengue_virus2.bam\n",
    "samtools sort $WORKDIR/dengue_virus2.bam -o $WORKDIR/dengue_virus2.sorted.bam\n",
    "samtools index $WORKDIR/dengue_virus2.sorted.bam\n",
    "\n",
    "jbrowse add-track $WORKDIR/dengue_virus2.sorted.bam --out $APACHE_ROOT/jbrowse2 --load copy --force\n",
    "\n",
    "bowtie2-build $WORKDIR/dv_genome.fa $WORKDIR/dv_genome\n",
    "echo \"Aligning comparison genome to dengue virus 3 genome using Bowtie2...\"\n",
    "bowtie2 -x $WORKDIR/dv_genome -f $WORKDIR/dengue_virus3.fa -S $WORKDIR/dengue_virus3.sam --very-sensitive\n",
    "samtools view -bS $WORKDIR/dengue_virus3.sam > $WORKDIR/dengue_virus3.bam\n",
    "samtools sort $WORKDIR/dengue_virus3.bam -o $WORKDIR/dengue_virus3.sorted.bam\n",
    "samtools index $WORKDIR/dengue_virus3.sorted.bam\n",
    "\n",
    "jbrowse add-track $WORKDIR/dengue_virus3.sorted.bam --out $APACHE_ROOT/jbrowse2 --load copy --force\n",
    "\n",
    "\n",
    "echo \"Comparison of dengue virus alignments successfully added to JBrowse.\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
