{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4a5a280f-ffde-440c-9181-c2debf66fd76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence ID: PP945018.1\n",
      "Description: PP945018.1 Flaviviridae sp. strain China-NX085, complete genome\n",
      "Sequence:\n",
      "TGTTGAGCGTTCCTTATCGGTGGTGTTTTCTCTGCGTTCCCTACTTTTTCGTTTGTTCACGTGTTCCGGCCGTCTTCACTTTAGACGCTAGCTCATTTCTTAATCGAGCTCGCGGCCTTTACTTACTACTGCTCCCGCCTTTAGGCGGGGGCCTCACACTCCAGACTTCTTGTGGTATTTGTTCTAACCCACTTTCCAACAAGATGTATACTCTTATCACAATCTCATCACTTTACTTCCACTTCTCTACTGCCCTTTTGGGCAACAACTCCGCAGCGTTTTGGTTTCCAGACGCTGACGGCCTCGACATGCCCCGCTGGCTAGCCCAGCAAGGGTACAACTACTCGTTTCACGGAGTGTGGAACATCCCCTCCTCCCATCACCCAGACTGTGAGCCCGAAGGGTTCGACAGTTGGGTCAAGGACAGGTGGCGTGGGCTTCTGTGGTTTGACAGCTCCCCAGAGACCAACCACACGTGTTGCCGCCTCCAGCACCACGAATACCTGAAGCATGCCTGTCGCGCTCAGGTGTCCCTGGAAGACTACCTGGCCGTCATCACGCACGCCCCCACCCCGGTCACGTCCCCCTTCTGCGCACGCCAGTGCCGCTTTCACGAAGCGGTCAACGTTGTGCAGGAAGCCGTTGATGCTGGCAGCTGGCACCCGTCGGACATCGTTGGGTGTAAAAAAGACGCGTGGTATCGCTTCGGCGACCACCTCGTTCCCCAGCCCTGCCTTCCCGACCCCCGCCCGACCATCGGCCCCACCCACCTCCCGCTGGGCCCCACTGACACGGGCCCCGGCATGGCCAAGGTCGCCCTCGAAAGCGTGACCGGCGTTGCCGCAGTCGCCGCGGTTGTGGCTCTGCTCTTCGTCCTCCGAACCAAGAGGCGCGCGGCGACTGCCCTGGCCCTCCTCCCCCTCGTCACCGACGCCGACGCCATCTGGACCACCAAAACCGTCGGCACCAACATGGTGGACGACACTGGAGCCCGTGACCCCGCCAACGTCACCATCGACTCCCTGGATTGGCACTGGTGGCCTCAGTGCTGCATTCGGATTGACGACCAGGCCGCCCAAACCGACCTCACCGGCGACAACTGGATTGCAACCCACGTTGCATCCGTTGACAGCCAGGCCTCCTCCCCGATCCACGACGACCTCCCCATGGAAGTTGAACTAGCTGAGCGCCCAGCCGAGACCAACTGGGCCCTCGACCCCACCACCCTGGCCCCCCTCTCTCCCGACGTTGCCTACCAAGGCGAAGGACTGGCCCTCTCCGCCATCGTGCCCGAGAGCATCGTGCACCGCGGCCAGCGGCAAAAGGAGAAGGGCACGAGTCCAGCCCCACGTCCGTCTGGCGCGCAGTCCTGCTCAGCAGCAGGCAAGAAGGCCCCCCCACCCGATGACGACGACGACTCCGACGCCGGCTCGACCTCTCCGTCCCCTCACGCCTCACCTCGCCCCCAAAAAACGTGCAGGGATGAAGCCGAAGGCCCTCGCTCCGGTAATAATGAGCAGGGCGGAGGCGGTGGAGGCTCGTCTGGCGGCGGCCCCACCGACACTGACACCAGCGGCCCGGCTGCGCATACCCGCAGCCGCAACAAAACCGGCAAGGCCCCCTCCGTCACCCCTGGACGCGGCCAGGCGCCACGACAGCGCCACAGCCTGTGGCCCTTCTCAGGGCCCCAGCATGACCGGACAGCCCGCCCCGCAAGCCTCGTGCTTGCTCTACTCACAGCCAAGAAGCCCAACGCTGGCACTTTTCTTTTTGTTGCCGCTTTCCTGGCAACTTTTTTTCAACCGACGTCGGCACACGGGTGCCAGCAGCTCATATGCCGCAACGCGGCTGTGTGGGTCTGCGAAGACACGACTTCCTTTGAGGTCGACACCCCTTCCGTCCACTGGTCCCGAGCGTGGCGGGAAGTCACTGGCTTCTCGTACCTCGACTGTGATCGCCGACTCACCGCCTCCTCCTCGCTTGCTGCCGCCATTGACGGCGAAGTCAGCAAGTGTGTGGCCCGTGGTCTCCGAGCCATCGGCCACTTTAACTCAAAGGGATACTTTGTCATCTGCATCCCGAGGCACGACCGTGACCCGACCCCCGCGCGCCACCGATCCAGGCGTGCCATCTCCGACTGGACTGACGGAGTGGCGGAGAAGGCCGGTGAGGCTTACAAAAAACTTCACGCCAAGGCGACCGACGCCGGCCTTGCGCTCAAGCGCCACGCTGGCGCTCACTGGGACACCCTCTACCACGTGGGGACGTTTCTCCTGTTTACAGGAGGGTGTCTCTTCGTCGGCATCCCCCCGACCCCCGCCATCCTCGCCGGCCTCGTCGCCGTAGCTCTCGCAGAAGAGCGCGACGGCACTACCCCGCCTCTCCCCAAAGCAGGCGCCACCTGCTACATCATCAGAGATGGCGCCCAAGTTGTCAAGCGCGGCGACACCGGCTGCCTCAACTGGCTGCGTGCCCAGGACCAGGCCTGTATCAACGCCATCCTGTCGTACCACCACGACAGGAAGCACGTCATCTGTGCCGGCCAAACCGCCACCATCTCGGAGCTCATCGCAGACATCAGCGATGTAGTCTCAGGTGAGGGGGCGTCTGCACTAAGAGCGGTAGGCAAGGCCACCCTCAACGTGACCAAAACGGCCCTTCTCAGGGCCAAAGGAAAAATACTTAAAACAAAAAATTTTTTTGAACGCCACCTCAACGCCAGTGCCAGCGGGGGACACACAGTCCTGCGCTGCGCAGCGACCGCGGCTGTGCTGCTCCTCGGAGCCCACCTGCGCCTCCCGCCACTCATCATAGTGGTGATCGCGCTCACCGCCTACGCCACCGTTGGCGGAGCCACTTGTGTCCACACCATCGGGCACAGCAGTACCGCCAGCCGCTACTACGACGACGACTACCACGTGAACCTCACCAGTGGTGAGGTCAGCGTCTCCCCCGGCACCTGCATTTCATGGTACCGCGAAGTCGGCAACGGTTCCAAGGTGCCGTGTGAACTGAAAATACTCAACGTAACCCAGACCTACAGCCGCGTGCCCATCTACCGGCTGGTCAAAAACTTTGAACTCACCGACCTCTGCGACTACAAGTGCGGAGGCTCTTTTTTTAGTGGCCAAGACTTGGTCACGGAAAAATTAATTCTGGCCAAGACGCACCGCGCCGGTAGAGGCATCGGTGCACGTGACAGCTACCTAGGCGTGGGGTGTGGCTCCTGCGCCTTTGGTACATGGGTACGTCACCGCTGCTATGTCCACTGGGACGCTGGGAAGGAGTGCGTCCTGCACAAGATCTCGCCCGCCCTACTAAAAATCGACCTCTCCTACACCGAGGCCGGCCATGTGACCAAGACGCAGGTCACAGTTGGTGAGCAGAAGATCAGCGTCGCCTCCCATGTCACCCTCAGCCTCCCTCCCTCACAGCGGACCGCTAGGGCCCAGCACACCCGTTTCATCAACTGCGGGGCTGACTGGTATGAATACGGAAATGACTGGCAGCTCCCGACCGACATCATTGTTTACGCGGAGGGCACGACCCCCGACTACGAGATGATCGCCAAGAAGATCGAAGGCGGCAACGACTGCACGTACGTCTGCCAGCAGCCGTGCACCACCACCTGGCGAGGCGCCCTCGACACAGCGCAGAACGCACAACTGCGCTCTGTTAGCGCCCCAGTTGAACAAGAGGCAGTGGCAGTTGCCAACATTGAACTTTCTGTTGAGGGAGTGCTCGTACAAGCGGGCCCGTGCACCCTCTCCCCCGACTCCACCGCCAACGCGACCATGGACCGGGACGGCTATGCACAAGCCGTCGTTACTTTCACTGCACAAGCCGAGGCTTGCGTGCTCACCCTCAGGGGCACGCAAAGCGACAAGTGCACCACTGCCGACATCGACGTAACTGTTGGCGTTGCCCAGACCTCTGTGCGCACCTCGGTCTTCTGCAAAGAAGCCGGCCCCTACCAACTCCTCGTCAATGACCAGAGCGTGACCGTCAACGTCCGCTCCGAAGAGGCCTGGTTCTACCAGAAGATATCCGGCCTCAAGGTCAATGTCGTTGGCCCCGGATACTCAAGCTGGTCTTGGCCCAGCTTTGAAGAACTCACCAAGAAACTAACTGGAAACTTCGCCGGGTTGGCTGACAAGGCTTTCCTTGTAATTTTTGCCGCATTTGTTGTCTGGGTGCTTGCACTCAACAATCCAATACTCGCTGTTCTCGTGGCGGCAGCGGCCTATTTTTTCATTTCTGCTAATGCCGAAGAAACTGACCCTGTCGCCACCGCCGCCGTCATCCTCGTGCGTGTGGTACAGGAGACCAGCGACCCCTACGTCGTGGCCACCCTCTGGGCATGGCTCGAAGCGTCCCCGGCCATCCTACTCCTCGCCGTCCACCTACTAGTTGCCGACCGACTGTGGCACCGGATGGCACAACTCAAGGCCAGCGGGGCCTTCGGAACACCGACCTGGCGAGTGACCAGGCAACTGCACACAACCGTGGGCAATTCACGAGTTGTTGTGCCCGCACTCTCGGCCTACGCTTTTTATGGAGGGCACGCCGCCGAAGGCTTCGCCCTCCTTGCTGTGAGCATGCTCCAACAGGGCGTTGGCTTCCTCCACTTCGGCAAGCACGTCGTGCTTTGCCTGCATACCCTCTTCCCAGAGAGGTATGCCAACACGCCAGACGTCATCGCAAACCTGCGAGACCTCTCCGACCTGTCTGGCGGAAAGAGCGGCACTTGCTTTGACAAGTGCACCCAGACCAACCCCAACGACGACGAATTCCGTTGGTGTTGTGCCGGCCGCAGCGGTACCCTGAACGAGCGCTGGACGCTCGTCAATGAGACCGTCTGCACAACAGGCGGCACCTACTGTGGCTTTACCGTCGACCCCAAGTGGACGACGAAGAAGAACGTCATCTATGACGAAGCCACCGAGCGCGAGACAGCCCGCCGCATCGCCATGGCTGGCATCTATGGCGAAGATGAGCTCCAACCCGGCGGGGACGAGGACACCCTCGCAGTCCGAGGAGCCCTGGCCGCCTTCACCACCAGGAAGAGCGCGCCTTACTTGGGAGCCAGCGTTGAGATTGGCTTCCTCCGTGCCATCGCCGCCAAGATCGGCTTTGACGGCTTCTTCAAAGAAGCCGACATCAACTTTGCAGCTCCTGCCATCTTCCAGATGCAGACCAACACGGGCTGCACGAGCACTGGCTTCTGCTTCCAGCACCGCGGAGCAGTGTTCACAAACTACCATGTGACAGGAGGCCGACCTCTTGTCCTCCCGACCGTGTCCAACGAAGGCAAGGTGGATTACTCGCTAATCCGCTGCGTCTTCAAGAACGAAAAAGACGATCTGGCTATCTACGGAAAGTCAGTCGCACTCGAGACCCCCGACATCGGAGAGGTTGGCATTGTCGTCAACCCCGCAGCACAACAGGCCATTGTGCTGCGCGTGAAAGGACCAGCAGCAGGCCACGAGACAACGCACAGCCGGCTAGACCGGTTTGTGTTCACCCAGAATTACAACGGGCAAGTGCAGGCCGTTGAGTACAACAAGTACTCCCTGTACGGGTGGAGCGGCCTGCCCATCATCTCCATGTCGTCCGGCCAGCCCATTGGCATTTACGGCAACATCGGAGTCGAGGATGGCAGGAAGAATGTCATGCTCAGCCCCCAGACCAGCGGGCTCGTCCGCACAGTGGACTGGCGCAATCAGGCGCGCCAGCTCCTCGACGGCCCTCTGGAAATCCAGAATATCGTCACCCCCACCGCCAGCGGCAAGAGCACGCTCTTCCCCCTCGAGCTGGCCATGGAGATGGTTGATCGCAAGAACTACGGCACCATCATCGTCCTCAACCCCCTCCGCGCGACCACCAGTGGGCTGGCAGCCTACGTCCGCGGACTTGTCACGGCCCGAGGCCTGGCAAGTAAAATCACTGTGGAATCCGCCATCGGCGGCGAGAGGAGCTTCAGCTCCATGGACTCAATCGACCGCGAGGCCAAGCAGAAGGTCGTCGTAGTCTACATGAGCTACGGCTCATACTGGAACATGGGGCCGTTCAACAGCACACGAGTGGTGCTCATGGACGAGTTCCACACACAAACTGTGGATGTCATGATCTGCCACGAGCGCGAGGTGTTCCGCCTGAACAGGCGGTTCCTCTGCGCCACCATGACGGCCACCGGCGCAGTCATGACGGCTGACCGCCCCATCGCCGACCACGAAGTTGGCGCACCGACGGCCGACCAGACCAGCGACTGGGCAGAGCTCACCGGAGGGCTCTATTTCCATCTGCCCACCTTCCGCGAAGGCCGGAAGCATCTGGTCTTCTGCGCCACCAAAAATGACTGTGAGAGCGTAGCAAGAGTGCTGCGTGAGAAGCTCCGCACCAAAATGGTGTACACATTCTACCGTGACCACCCGGCGGGAGCTTCCTCCATCCAGGCCTTCAACAGCAGCAGGAAGGCCATCATCGTCGCCACCAATGCCATCGAGAGTGGCATCACCCTGAATGACCTGACAGACGTCTGGGATACTCAGGAGGAGAACGTTGTCGAAGTCGAAGACAACGGCAACACCTGCTTCAGGAGACGACCGATCTCCAAAAAGAGCGCTATCCAACGGCGCGGACGCTGCGGTAGGACCACCGACACCGCAAACTACTGGTTCCCAAGGGGCAACGCCCTGCCCACGGATCAACCAATCGACTACGAGATCAAGCTGCTCTGCGCAGCTCACTGGCTCTCAGTCTCAAATGAGCCAGCATGCAGCCCCGAAACGCGCGAAGTTATTGACAGGCTTCGTGCCATGTTTGACCACCATGGAGTCCACCAGGGCTTTTGGCGCCTCCTCACGTTAGCTTGCGTGGGCCCGAGCTCTTTCGCGGCCGCAGCCTTCAGAGATGCCCCCCCCCGACCACGATACAGCGAGACGTTCCTTCGGGCCGTCTTCGCGCGCGCGATCAAGGAACAGCACACCTGCTGGAAGACGATCCGCGCAAGTTTCATAGTCCACCCAGAGCCGCCATCCGTCCTTCCCGAGGGCGTGGAGCTGGCTCAAATTTCCATGATCCTGAACGCTAAAGGCGAGGGCGAAGCCCAAGAAGAGAAAAACTCAATGACCGCCATGCTTGGCACATTGGGTTTGGCGGCAGCCACCGCCACAGCGCTGGGAGAGCTGGTGGCAAGCGTGAGAAGCTTTGCCACGCCAGTGAACATGTGGATCCTCCCCAGCGACAACTGGCTGGACCTGGCATGGGGATGCCAGTATGATCTTGTCCTGGGCTACAGCCGCTCAGACTCAGAAATCGCGGCTGGCCACCTCGATCCGCACATCCTGGCCACCAACCCCCTCTTCAAGAGGGTCCTCAGCGTGGACTCCCCCATCCGCAGCAGTTGCCACTATGCTGTTCACGGCGTGAACAGTGGCGCTCTGCGCTACTTAAGACAAGTCTCCGTCGGCGTAATGGGAGTCTTTGTCTACAGCACCATCTTCGACAAAGCCGACGTCGCTGCTGCCGCTGCAGCGCGACACATCGAACCCAGCTCGCTCGTAAGCCAGCACGAGCAGAATGCTGTTACTGTCAGAGGCGCAGGCTGGTTCGACATTTTCTGCGCCTGGGTCAAGTGCCACGACACCACCCCCCCTCTCGACGGATGGGAGGAGGCCGAGTACGAGGCCCAGTGGAGGATCAACCACGGCAGCGAGGTGGTGCAAGATCACTACCGCACCGCGTACCATCGCGTGAAGGAGCAGCTTGCAGCCGTCGGCGTCGAGCAGCGCTCGGAGGATGAAGGCCCAGAAGCTCCAGCTCACGGAGAAAGCGCACTCACCACAGGCGCCGTTTTGGCCGCTGCCACGAAAGCAGCCACAGCCTTCTCCACGGCAGCCTCTTCCGCTGCTACCACCGCTATGGCTGCGGCCCCTGGAGTTGCCGCATCAGCAACCGCGGCTTTCACCTCAGCCGCCTCAGCCGCTACCGCCGCAGCCTCCACTGCTGCTACCGCCGCAATCGCCGCAGCCCCCGGCGTTGCCGCAACAACAGGCGCGGCACTAACGTCTGCCGCCACCACCGCAACAGCAGCAGCGACAACAGCTGCTGTTGCAACAAAAACAGCCATGGCAGGCTACGCCGCCCAGCTGAGCGCCTACCTCACGACACAGGGCGTCAGTTCAGCAACAGCCGCCACCCTTGTGGCTTGGTGGCAAGGCGGCGTCGTCGTCGCCGGAGGAATCCCACTCGTGTGGGACCAGCTAGTCGGCAGGCTGGGAGACCATGCCGCAGCCGGAGCAGCCGTTCTAGTGATGGCTATGTGCGCCAACCCTGGCTACCTACTGAACAACGTGTTGGCCAGCACCCTCACGTACGGCCTTAAGCACGTCGCCGTCGGCAACGCCGTAGTCAACGGCTACAGGGTGGGCACCACCCCCTACCTTCTCGCTGGCTCCGTTGCCCTCGGCACGGCTCTCGCCAGCATGCGCGTCGTCAACAGCGCAGACGCAGCCACCGCCCTGGCAAACACCATGTCGCCCAGCGGCGTTGTCCTCAGCAGCACCCACGGGAGTGCTGTCGTTGTGGTACAGCGGCTCTACCACATCTTCCGCCGGCTTGTCGAGAGCCCGGCCGACATGGCCAACATCTCTGATGTTGCCGCAAGCGCCATCACAGTCCTCGTCGAAATGGCAGGACTCGGGCCCACCCAAATGGTCACCATCGTTGTGGCCACAGCGACCCTTGCAGCCATAAGAACGGCTTGTAAGTATAATGAGCGCTTCCGAACGCTCCTTGGATACTACATCGCCGGCAAGAACGGCGAACAAGTTGCGGCACTTGAAGAAAATGCCGATATCGACTACTATATCGACCTGGCCCTCTCCATCGCGGCCCTCGTTGTGGCCCCCAGCACAAACATCAGTGTTGGAATCCTGCTCGGCTGGAATGCCGTGCTCGAGCTCCTTGAGGGGCGCAGCGTGACGAGACACAGCCTTCGAGGCCTTGTGTACGAATCCGTTACTGCAGGGGCGGGCCTCCCCATCTACATGGCCCTGGGCAGCGTCATCGCTCAAATCAGTGCCCGATATGCCAGTGGCCGCCTTCAAGCACCAACAGTTGAAGGCGAGCACAACGCTCTGACCGCCATGAGCGGCATTTTCGGCTGCTTCGGCGCCGTAGCCAAGGGTGTCCAAGTTGACATCCGCGAGGCCATCAACTCACATCGCTGGAGGCACTGCGTCATCGTGCGCTTCTTCCAGTCCTGCGCCGGTGGCATCGCCGCCTTTGGAGCATGCGTCGTTGAGGCCTTTCGCTGGGTCAGAGACCGCTTCTGTGCCCTCTTCGGCATTGACAACGTCCCCGGCTACGACGAGCTGGTCGAAAATGACGGCAGCGCAGACCTTTTCCGCATCGAGACCGCGGCCAAAGAACGCGACGAGGACATCATCTACAGCGAAGTCCTCATTCAGAAAGACGACGAAGAGGCTGTCAAACGCCTCGCTAGCGAGTGCGCCAACATCCGCCAGAACGCCATGGCCAGTCAGCACCTGGTGGGCATCTCAGCATGGCTCGACCTCGATGAGAGCATGTTTGAGAGCACCATTGTGTTTCAATATGTCGTCGAAGACGTTGGCGGCATCCACTTCGGGCTGCTGGCCGAATGGCTAAAAAAGACCTACCCAGACGTAGTTGTTCGTGGGGCAAGTGGTTACTCGAACCTTGGCTGTGAGTACACCACCCTCCTCACCGACGCAGGACCCCTCACGCTTGTCCACACCATCACGGACAGCGGCCTGGACGTGTACCAGTGCCTACACTTCAACAAGCGTGTCGTCACAGTGTACACCCTCAACACCCTGGGGTACACACAAGACCGTGGCTTTGACACCCACGTCAGACCCCGCCTCCAGGAAGTGCTTGAAAAAGCGGCCGAGGCGCTTGAGGCAGTTGGGTGCGGCAAATCAGAGCTCTTGCCGCACTTCGTCAACCGGAACGTCCTCAAGGACTTCCCCGCCCTCAAGAAAATCTATGACGAGGGAAGAAAGTGCCTCCCGGCACCAAAGAACAATATCGTCAAGCCTGCCCTCTCCATCGGCGGGGGCATCAAGATCGTTGTCAGCAGCATCTTTGGGGCCAGGTACCCAGTTCCAACCAGGTACCTTCAACACCTCGACGGCAATGACCACCTGTCAGCGAACCAGGTGGACCCCATCGCGACCGAGGCCGACTGCGAGCGCTTCCGACAGCGCCTCCTCATGAAGTCGTACGCCATCGGCTCCGCTAGTCCCGCATTCGATCTTGCGGACAGCATCACAGCCCTTCGGGGCATTCAAAAACAAGGACTGTCAAAGGTGGCCCTAAACCTCCTCATTGACTACCAGTGCGCTGCGCGCCGCCCAACCCTCGAGGAAGACGTCAGGCTGGGCAAACCACACAAGCTCGAGTACATCCGGGCCCTCATCCACGAGCACAGGCGCACAGGCGCTGTCGGCGAAGTCCACGCCGAAAACAACTGCGACAGAGAGGGAGGCAACATTCCACCTTCTGTCGAACCACGACCACAACCAGTGGCGTACAAAGTGCGTCATAAAAAAGCCTCTAATGTCATCGGGGGCAAAGAAACAACAGCGCGCGCCATTGTCAACGAATGCCGCAAGCGCGCCGCCGCTGCCCAGTGCTCCTCAGTGTACATCGTGCACGAGCGCCTTGCCTCCGTGAGGGGCAACCTGGGCAACCTCACCTACGACGACTCACCGCTCCCACTCCCAGTGCCCATCGGGTCATGGGACAAGAGTGAAATCGACACATACCAGCAGGTGGAATCCACCAACGCCATCGCTCGGTATGTCGCTGACATCACCCAAGGCTCCGAGCTCATCATCGAACCATGCGCAGGTCTGGGCCGGCTAACAGCCCTCTGCGCTTACCAGAACAAGACACCGCTCTGGCTCTCTGACATCTCGGTGCACTGCGTCGAGCAGCTCAAGCACCTCTACCAACGCCCAGACTGCCCCGCCAAGGTGGTCATCGAGCAACGAGATGTCATGAGCCTGGGCAAACAGCTCAAAGGCCAGACGGTGGTTTGCAACCCGCCATACAACCGACCCGAAGACGCCACCCTTGGCACTAGGGTCTGCAAAGCCATCTTCAGGAACTGCCCTGCGCGGTTCATCTGTGTCTGGCCCGCACGACTACACGAGCAGTACATCAGGCTCGTCCCCAAAGAGTACGACACGGCAGCCAAGCCGTTCGTACTTACCATCCCTGGGGTCCTGCCACGTCACCGGCAGAAAGAGGCGTACCTCCCCGTGTACGTCTACGACAGCATCCGACGGGGCACAGGCGACCGCACCGAGCCCAGGGTGCAATTCAACAGCGCCCTCGCCGCAGTCGAGGAAGAGCTCAGACAGCTGGACGTCATCACCAGCACCATCGCTCCCCAAACAGTTGTTGACACCATCAACAGCTGCCAGCACCCTCTGGGTGCCAAATTTGATCCTGCACGCAAAAATGTGCCATTCTTCAACAACAGGCAGGACAAACCTTTCATGGCAGCCCCCGTCAGCACACTCCTTCTCCACAAAAACATCGTGGAGGAGATGGCAGAAAACCCTACCGGCAGACGCGCAGGCCGGTGCATGCCTGCAGACCCCAAAATCGGCGCTGACTGGGCCCTCGCCGCGGACGCCGACAAGACCTTCCAACTCCTCGTTGGTTACCGCGTGGCCAACGACGGCAGCATTGAATGTATCGGTAAGAAGACCCACACCGACATTCTTTGCCTCGGCCTTCCCAAAGGCCAAGTCGACTACATCGCGCAGGCAGCAAACGCTGCCACCATACAAACTGACGTCACCGCACACACCATTGCGGCAGGCAAGACTAACGTCGTAGACCGGCCGGAGAACGTCAACCATCCGGCCTCTCTTGTCTACGTTAACCTCGACGTCATCCAAGACACGAGTGGCATCCAAACAACGCCGCTCCAAACACGAGTCTACTACACCGAGCAGCCCTGGAACCTCAGTTTCTGGCAAGAAGTCGGACGCTTTCCAGGGCTCCGCATCCATTGGTCACCGACATCGCCCAATGGAACCTTTAGAGTGTTCGTGGAAACACGAGGACACCCGCGGCGCAAAGGCATCAATGCCCGCTACCTCTGGAGCTTTGCGGTGCGGCACATGCTCGACGCAGTGTGCCGTCAGCTCGAAGACATTCACACAGTCAGCCCGAAGACCTGGTTCAGCTACCCCAAGTTCGAAGACCACCTCAAGCAGTGGTCACTTGTCAAGAGCGAAGAAGCAGAACACCAAGACTTCTCGCGCCAAGCAAACCGTGCGAAAGTGCGGCAACCAGTCGAGCTCAAGGGCGGCTGGCCAGGCTTCGAAGGCCTCAAGGGCCAATTCAAGAACAGGCTAGACCGACTCGACGAAACTTGCGCTCACTACAAGATCAAGCTTCAACCCGCAAGCCGAGCGCGCGTTTTCCAAAACCAAGAGGTTTTGGGGACGTACAACAAAAATGTCCGACTACAAGAGGACAAGACAATGCTGCAGCAAACAACATACCGCGCAGCAAACAGCATCTGGGAGTGGGACCCGCGCTGCTCCGCCGTCAAACAAGTGAGCAGAGACCGCGTGGAGGTCCTCGGAGCGATCCAAGACCGCATAGACGCTCCACATTACTTCCCAGCCCCCGACGTCCTGGAGGAACTGCACACCATCGCAGGCCACATCCGGACTAACATGGCCATCAAACTTAAACCTTTCACATTCGAGGAGGCGCGCACCATTATCAATAAACAAAGTGCCTCTGGAATGTTGGACGACTGGCCTAACATGGAGGCTTTCCTCAATGACCCTCGCGCAAAAGACATCTGCGAGGAAGCCATCCAAAAACTATACCGAGGAGAAAACGTTGTGGAATATTACAACACCGTTCACAACAAAGTCGAATCCAAGGTCAGCGAGAACACCCGCCCCCGTCTCATCAACTACGGGAGCATTGTGGCACGCGTCTGCGACCACATGCTCTTCGGGCCCTTCGTAAAGCTTCACTACGGTGGCACGAAACTTTTCAAACACGCCAGCGGAGGGACCCCAATTGCGGAGATGGGCGACCGCATCAAGAAGAACTGGGATCACTGGGCTTCCAAGTGCGTGGGAGACGAAGAACCAGTCGCTGCTGTTGGCGACGCCTCGCGGTGGGACCACAGCATGTACCCCGCTCTCATGGCACTGGAGGCCGAGTTCGTTTCCAGTTTCTATGACAAGAAGCACTGGCCGGCAATCCAGGCAGCCCTAGAACATGTCTGCTGGCCAATTTGTTTTACGAGGCTTGGCTTCGTTTTCACGAACCCAGGTCAGCGCATGAGCGGACACGTGTTCACCTGGATGAACTCGTTCCTCAACGCTGTCCTGCACGAGTGGGGGTGGAAGAAGACCCTGAACATCCCGCTCGACGCCAACCTCGACGACTACTTGACATTCGAAGTCGATGGAGACGACAACTACCACATCTCCACCACCAGCGTCATTAACGCGGAATCCATGGAGCAGCTGCGCTTGAACATGGCCAGCGTGAACATCAAAATCCGCTCCAAGACGTTCGACGGATACCGCATCACCGACAGAGTAGATGACATCGACTACCTGTCTCATTACTACGCACCGACTCTCATCAGGTGCGAAGCACACGACTCCGGAGCGCGCCCTCCGACCACAGACACTGGCGTCACGAGCACTTGGCACAAAGAAAAATACCTGCCAGTGCGCCCCATTGATGAAATCTTGGGGAAATGGTGCTTTACCATCAAACACAGCACCACACGTGACTCCGCATACACGTCACTCAAAGACGTCTGCGGTCAGGAGGTGTCCCTCCTAAACAAAACTGGACTTGACATTGAGAGCAGCAAGGCCCTAAGCTACCTGCTCCAATACCCCCACATCCGCACGGTCCGAGTCATGGCCATCACCATCCTCAGCCAGACCGGCTTTGCCAGCAACCCCTCGTGGTTCAAAGCCTACCGCACGCGAATGAAACTTGGCGAAGGGTGCGACGCGCTAGATTCTGCGTACGAAGAACGCGCAGCAGACGCCCGCGCCCGCGAGGCCACCATCATCACCAACGACGACTTACTAGAGGCCGACTTCCATCTCGAGGCCGAAGCAGCAACCGGCATCGTCGTCGGCGCACTCCGGTCGATCTATGGCCAATCCATCCGCAGCCTGGACGACATCGGGTGCGTCGACTTTCGTTGGGAGAGGAGGAGCTGCCGCGAAATGATCACTCTCATGGTCAACGGCGCAAAGGCTGGCGCAGGCAACTCTGTCACACCGGACATGGAACGTTGTGCACTGGACGCCAAGTTCAGCCACAAGTGGTGCAAAACCACAAACCAAATCATCCGAGCGCTCCACCACCTGCGCTCCAGGTGTGAGCAGTTCGGCATAGCTGCCCCGCCTTTCCCACACTTTGCCACCCTCTGGAAGCCCTTCCTCTTTCACTACTCGCGCTTCTTCAAGACGCGAAATATCTACACAACGCCGCTCCCCGAGGCCCCCAAAAACATCATCGATGACTTTGCCCAACGAGCAATGGGCTTAGTCAAGCAAGCCCCAGGCTACGCCATTAAGAACGTTGGCGACATACTGGGCTGGCTCAGGCGCACTGCTCCCGACTCATTCGCTCGCCACAACAACGGTGCTGCGGATGGAAACCATGATGGCGGCGCCACAGCTGATTACCGAAGCGAGGGAGGCAGCAAGCGCGGTGGCTTCGGCGGAGGCGGTGGCGCCGGCGGCGGCCGCGGTGAACTCAAAAGAGGAGAGGGCCAAACGAGGGACGCCGGCCCCGCAGGCGGCGTTGGTCCCAGCGACAGTCAAAACCTCTTCGACAGAGGCGTCGTGGCTGTTCGTGGCATCCTCGGTAGCTGGCTCAGGCCACGCTACCACACCGAAGAAGACAACTTGCCAACCTGGGAGGTCATCAAAAAGGCTTCCAAAGCGACTATCGCCTCCAGGAAGGCAGGTGGGGGCAGGCGAGGAGTGCGCCCCCAAAGAAACTGTGACGAAGACTCAGGCCGGCCCTTCAACAACACTCGGGCAATCACGGCTACTATCTTCTTCACCATCGCGGCCATCGCTGCGTTCTCAGTTGGCAGGAAAGCCGTCGGAGTCAGCAGCACCATTGCCAAGAGAGTTGGCTGGTGGGTGACCGCGTTCGCCGCCTTCTACGGCACTACTACAGCCGCTCACTACGTTGCCGGCAAACACAGAGGGCCTGGACAATCCCTCCTACCTCGTGACTGGCTCAAAAGAGCACGAGCTAGAAAAGTCCAGGGTGGAAAGCCCGCCCTCGAATACTCAGGAAACGTTCCCTTCTGGCAACAAAGAGCAGGGGGCGAGCGTGGGCGGCACCGCAGAAGCAGAGGCCTTTGCAAGATCAAAGAGGCACTCAAACTAGCCCTCAGACTTGCCTGGCGCATTGCCATCAAACTACTAGTCATTGGCCTCGGCTTCTGCGTAACGCTTTTTTCTATTTTTTATCCATTCATTGTACCTTCTTTTCTTTTTGTTTTTATGTGCATAATCTACTGTTCCAGGAACGCACGGCTCGCCCTGCTCTGGTACCTACTCGTCTCACCGTCCTTTGACGCGGCCGCTTTGGCCGTACGAACAAAGGTGGGAGACTCTGCTGCGCAATTTGGCGCTATACCAGCCATTCTGCGCGCATATGTAAATTCTTAAAGTCAAGTACGTGATGTCCACACACACACACACACGCGAGGCCGTAAACCTCGCCACACCTGGGTAACCTGCAAGCCAGTGCACTAACGTAATCAGACGAGTGATGCTGCACCCAGGCATACACAAACAAAACCGAGCTAAAAGCTCGCCCCGGGGTAAAGTCTAAGTTGTAGTACTAACGTAAAGCCGAGTATCTGCACACCCCGGACGCCCGCATGGGCACCCAAGTGACAGGTTATAAGGCCGCTTGATCGAGCAATGATCACCGGCACCTGCCCCAAGAGTTTAGGCCTCTTGTTAAAAACCTAACCCTAAAAACTTCGACGGACCATTAAGCACCAATGGCGAATCCCAATTGGCAACTGCCTTGGATGCTGTCCGTCAACCCCCAAAATTTTGAGGCCTGTCGTGCGAGGGTATTCCGCACCACAGGTCTCCTGTCCCAAAATTAAAACCAGGTCAAAGTATTGTATCTTCTGATTTGAGTGGGATTTGGAGGAGTCTGGGACCCGTTTTTTGTCATTAAGCCTGACAGTAGCAACGTATCTAATACAAACGCGAACGCGATGGCGACATTGACGAAAACAAAGGCGGGTCGACACAACTCCAACTTTTCTTTCACTCAACAACCAGGGGATCCAATACCGGAGATC\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Set your email (required by NCBI)\n",
    "Entrez.email = \"smrithisurender@berkely.edu\"\n",
    "\n",
    "# NCBI link or Accession ID\n",
    "ncbi_link = \"https://www.ncbi.nlm.nih.gov/nuccore/PP945018.1?report=genbank\"  # Replace with your link\n",
    "accession_id = ncbi_link.split('/')[-1].split('?')[0]  # Extract clean accession ID (e.g., PP945018.1)\n",
    "\n",
    "# Fetch the sequence\n",
    "handle = Entrez.efetch(db=\"nucleotide\", id=accession_id, rettype=\"fasta\", retmode=\"text\")\n",
    "record = SeqIO.read(handle, \"fasta\")\n",
    "handle.close()\n",
    "\n",
    "# Display the sequence in the notebook\n",
    "print(f\"Sequence ID: {record.id}\")\n",
    "print(f\"Description: {record.description}\")\n",
    "print(f\"Sequence:\\n{record.seq}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a5133275-d05e-4bd2-b430-d9102638f27d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error fetching from https://www.ncbi.nlm.nih.gov/nuccore/advanced: HTTP Error 400: Bad Request\n",
      "Error fetching from https://www.ncbi.nlm.nih.gov/sites/myncbi/nuccore/filters: HTTP Error 400: Bad Request\n",
      "Error fetching from https://www.ncbi.nlm.nih.gov/sites/myncbi/nuccore/filters: HTTP Error 400: Bad Request\n",
      "Error fetching from https://www.ncbi.nlm.nih.gov/nuccore/details?querykey=1: HTTP Error 400: Bad Request\n",
      "Table Size: (80, 3)\n"
     ]
    }
   ],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "from Bio import Entrez, SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "# Set your email (required by NCBI)\n",
    "Entrez.email = \"smrithisurender@berkely.edu\"\n",
    "\n",
    "# Main NCBI page URL (replace with your page URL)\n",
    "main_page_url = \"https://www.ncbi.nlm.nih.gov/nuccore?term=%28%28%22Flaviviridae%22%5BOrganism%5D%20OR%20%28%22Flaviviridae%22%5BOrganism%5D%20OR%20Flaviviridae%5BAll%20Fields%5D%29%29%20AND%20%28complete%5BAll%20Fields%5D%20AND%20genome%5BAll%20Fields%5D%29%29%20AND%20%28%22Homo%20sapiens%22%5BOrganism%5D%20OR%20human%5BAll%20Fields%5D%29%20&cmd=DetailsSearch\"\n",
    "\n",
    "# Function to extract genome links from the main page\n",
    "def extract_genome_links(url):\n",
    "    response = requests.get(url)\n",
    "    soup = BeautifulSoup(response.text, \"html.parser\")\n",
    "    \n",
    "    # Find all links on the page that lead to genome records\n",
    "    links = [\n",
    "        \"https://www.ncbi.nlm.nih.gov\" + a[\"href\"]\n",
    "        for a in soup.find_all(\"a\", href=True)\n",
    "        if \"/nuccore/\" in a[\"href\"]\n",
    "    ]\n",
    "    return links\n",
    "\n",
    "# Fetch sequences for all extracted links and create a table\n",
    "def fetch_sequences_table(links):\n",
    "    data = []  # List to store rows of the table\n",
    "    for link in links:\n",
    "        try:\n",
    "            # Extract accession ID from the link\n",
    "            accession_id = link.split('/')[-1].split('?')[0]\n",
    "            \n",
    "            # Fetch the sequence\n",
    "            handle = Entrez.efetch(db=\"nucleotide\", id=accession_id, rettype=\"fasta\", retmode=\"text\")\n",
    "            record = SeqIO.read(handle, \"fasta\")\n",
    "            handle.close()\n",
    "            \n",
    "            # Add a row to the table\n",
    "            data.append({\n",
    "                \"Sequence ID\": record.id,\n",
    "                \"Description\": record.description,\n",
    "                \"Sequence\": str(record.seq)\n",
    "            })\n",
    "        except Exception as e:\n",
    "            print(f\"Error fetching from {link}: {e}\")\n",
    "    # Convert the list of rows to a DataFrame\n",
    "    return pd.DataFrame(data)\n",
    "\n",
    "# Extract links from the main page\n",
    "genome_links = extract_genome_links(main_page_url)\n",
    "\n",
    "# Fetch sequences and create a table\n",
    "sequences_table = fetch_sequences_table(genome_links)\n",
    "\n",
    "# Print the table size\n",
    "print(f\"Table Size: {sequences_table.shape}\")  # Number of rows and columns\n",
    "\n",
    "sequences_table.to_csv(\"genome_sequences.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99ca23db-7411-4676-9ac6-9163dda97ec5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f2fd3c74-3c6c-4ba0-b090-270eab1c0487",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error fetching from https://www.ncbi.nlm.nih.gov/sites/myncbi/nuccore/filters: HTTP Error 400: Bad Request\n",
      "Error fetching from https://www.ncbi.nlm.nih.gov/nuccore/details?querykey=1: HTTP Error 400: Bad Request\n",
      "Table Size: (80, 3)\n"
     ]
    }
   ],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "from Bio import Entrez, SeqIO\n",
    "import pandas as pd\n",
    "import time\n",
    "\n",
    "# Set your email (required by NCBI)\n",
    "Entrez.email = \"smrithisurender@berkely.edu\"\n",
    "\n",
    "# Refined NCBI search URL (example, specify organism and genome)\n",
    "main_page_url = \"https://www.ncbi.nlm.nih.gov/nuccore?term=(Flaviviridae[Organism]%20AND%20complete[All%20Fields]%20AND%20genome[All%20Fields])%20AND%20(%22Homo%20sapiens%22[Organism]%20OR%20human[All%20Fields])&cmd=DetailsSearch\"\n",
    "\n",
    "# Function to extract genome links from a single page\n",
    "def extract_genome_links(url):\n",
    "    response = requests.get(url)\n",
    "    soup = BeautifulSoup(response.text, \"html.parser\")\n",
    "    \n",
    "    # Find all links on the page that lead to genome records\n",
    "    links = [\n",
    "        \"https://www.ncbi.nlm.nih.gov\" + a[\"href\"]\n",
    "        for a in soup.find_all(\"a\", href=True)\n",
    "        if \"/nuccore/\" in a[\"href\"] and \"advanced\" not in a[\"href\"]\n",
    "    ]\n",
    "    return links\n",
    "\n",
    "# Function to check and extract next page URL (pagination)\n",
    "def get_next_page_url(soup):\n",
    "    next_button = soup.find(\"a\", string=\"Next >\")\n",
    "    if next_button and \"href\" in next_button.attrs:\n",
    "        return \"https://www.ncbi.nlm.nih.gov\" + next_button['href']\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Function to handle pagination and get all links\n",
    "def get_all_genome_links(start_url):\n",
    "    all_links = []\n",
    "    current_url = start_url\n",
    "    \n",
    "    while current_url:\n",
    "        # Extract links from the current page\n",
    "        links = extract_genome_links(current_url)\n",
    "        all_links.extend(links)\n",
    "        \n",
    "        # Fetch next page URL\n",
    "        response = requests.get(current_url)\n",
    "        soup = BeautifulSoup(response.text, \"html.parser\")\n",
    "        current_url = get_next_page_url(soup)\n",
    "        \n",
    "        # To prevent overwhelming the server\n",
    "        time.sleep(1)\n",
    "    \n",
    "    return list(set(all_links))  # Remove duplicates, if any\n",
    "\n",
    "# Fetch sequences for all extracted links and create a table\n",
    "def fetch_sequences_table(links):\n",
    "    data = []  # List to store rows of the table\n",
    "    for link in links:\n",
    "        try:\n",
    "            # Extract accession ID from the link\n",
    "            accession_id = link.split('/')[-1].split('?')[0]\n",
    "            \n",
    "            # Fetch the sequence\n",
    "            handle = Entrez.efetch(db=\"nucleotide\", id=accession_id, rettype=\"fasta\", retmode=\"text\")\n",
    "            record = SeqIO.read(handle, \"fasta\")\n",
    "            handle.close()\n",
    "            \n",
    "            # Add a row to the table\n",
    "            data.append({\n",
    "                \"Sequence ID\": record.id,\n",
    "                \"Description\": record.description,\n",
    "                \"Sequence\": str(record.seq)\n",
    "            })\n",
    "        except Exception as e:\n",
    "            print(f\"Error fetching from {link}: {e}\")\n",
    "    # Convert the list of rows to a DataFrame\n",
    "    return pd.DataFrame(data)\n",
    "\n",
    "# Get all genome links from the refined search\n",
    "genome_links = get_all_genome_links(main_page_url)\n",
    "\n",
    "# Fetch sequences and create a table\n",
    "sequences_table = fetch_sequences_table(genome_links)\n",
    "\n",
    "# Print the table size\n",
    "print(f\"Table Size: {sequences_table.shape}\")  # Number of rows and columns\n",
    "\n",
    "# Optionally save the table to a CSV file\n",
    "sequences_table.to_csv(\"genome_sequences.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c25e771e-1521-4557-b77a-ce541b4a54f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GFF file saved as 'output.gff'.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/srv/conda/lib/python3.11/site-packages/Bio/SeqFeature.py:231: BiopythonDeprecationWarning: Please use .location.strand rather than .strand\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# Function to fetch GenBank file using NCBI Entrez\n",
    "def fetch_genbank_file(accession):\n",
    "    Entrez.email = \"smrithisurender@berkeley.edu\"  # Provide your email for NCBI access\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=accession, rettype=\"gb\", retmode=\"text\")\n",
    "    genbank_record = SeqIO.read(handle, \"genbank\")\n",
    "    handle.close()\n",
    "    return genbank_record\n",
    "\n",
    "# Function to convert GenBank annotation to GFF format\n",
    "def genbank_to_gff(genbank_record):\n",
    "    gff_data = []\n",
    "    \n",
    "    # Header line for GFF\n",
    "    gff_data.append(\"##gff-version 3\")\n",
    "    \n",
    "    # Extracting annotations from GenBank\n",
    "    for feature in genbank_record.features:\n",
    "        if feature.type in [\"gene\", \"CDS\", \"mRNA\"]:  # Focus on genes and CDS\n",
    "            # GFF format fields: seqid, source, type, start, end, score, strand, phase, attributes\n",
    "            seqid = genbank_record.id  # Use GenBank accession as seqid\n",
    "            source = \"GenBank\"  # The source of the annotation\n",
    "            feature_type = feature.type  # gene, CDS, mRNA\n",
    "            start = feature.location.start + 1  # GFF is 1-based, GenBank is 0-based\n",
    "            end = feature.location.end\n",
    "            score = \".\"\n",
    "            strand = \"+\" if feature.strand == 1 else \"-\"  # GFF uses + and - for strand\n",
    "            phase = \".\"  # Phase is used for CDS, but we can leave it as \".\" for gene or mRNA\n",
    "            attributes = f\"ID={feature.id};Name={feature.qualifiers.get('gene', [''])[0]}\"\n",
    "            \n",
    "            # GFF formatted line\n",
    "            gff_line = [seqid, source, feature_type, str(start), str(end), score, strand, phase, attributes]\n",
    "            gff_data.append(\"\\t\".join(gff_line))\n",
    "    \n",
    "    return gff_data\n",
    "\n",
    "# Function to save GFF data to a file\n",
    "def save_gff(gff_data, output_file=\"output.gff\"):\n",
    "    with open(output_file, \"w\") as f:\n",
    "        for line in gff_data:\n",
    "            f.write(line + \"\\n\")\n",
    "\n",
    "# Main execution\n",
    "def main():\n",
    "    accession = \"KX528231\"  # Example: SARS-CoV-2 genome (NCBI Accession ID)\n",
    "    \n",
    "    # Fetch GenBank file\n",
    "    genbank_record = fetch_genbank_file(accession)\n",
    "    \n",
    "    # Convert GenBank annotations to GFF format\n",
    "    gff_data = genbank_to_gff(genbank_record)\n",
    "    \n",
    "    # Save the GFF data to a file\n",
    "    save_gff(gff_data)\n",
    "\n",
    "    print(f\"GFF file saved as 'output.gff'.\")\n",
    "\n",
    "# Run the script\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aaebcba6-27f4-48c7-991c-1e9eef1dd8a6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
